### DISCUSSION TITLE
Parse: thru any vs. thru some

### DISCUSSION THREAD

**Discussion by gor77** [red/red#5524] [Q&A] [answered]
These first two seem to be ok: 

```
parse [1 bye bye] [some word!]
; == false 
; (ok, because it can't pass the beginning integer)
```

```
parse [1 bye bye] [thru some word!]
; == true
; (ok, because thru passes the beginning integer) 
```

But what about this one:

```
parse [1 bye bye] [thru any word!]
; == false

```
`Any `should be more inclusive than `some`.

**Comment by hiiamboris** [comment, marked as answer]
`any` succeeds on zero matches as well, so `thru` also succeeds without advancing the input, then you have 3 more tokens left, and `parse` returns `false` indicating that it hasn't consumed all input

It's better to use our matrix rooms for such specific discussions, e.g. [red/parse](https://matrix.to/#/%23red_parse%3Agitter.im?via=gitter.im&via=matrix.org&via=matrix.freyachat.eu)

There's also a [ParSEE visual tool](https://codeberg.org/hiiamboris/red-spaces/src/branch/master/programs/README.md#parsee-parsing-flow-visual-analysis-tool-parsee-red) available that aims to help understand the parsing progress, give an overview and let you explore it step by step (good for more complex rules or rule dictionaries). Or the low-level `parse-trace` wrapper available in the runtime (good for simple rules like this one).

**Reply by gor77** [reply]
That was one of the things I was afraid of - it's more complicated then it looks like :). On the surface, it looks
deceivingly simple...and there is abyss beneath.

So if I got things right, the key thing to understand about `thru `is that only *failure* of the subsequent rule 
moves the input position forward. Then it makes sense. Since `any `succeeds no matter what, `thru `is, in a way, left paralysed.
But it does succeed. `Any `is done, `thru `is done...with input position still at the beginning. 

Tnx for the tools, I'll check them.

**Comment by greggirwin** [comment]
Red is more protective of infinite loops, but you still have to be careful with combinations of things that may not advance the input.

**Reply by gor77** [reply]
There is something else here I completely forgot, and that is the terminology used in the docs.

For example, when it comes to `any `:

From the official blog post:
`any `_rule_	: repeat _rule_ zero or more times until failure or if input does not advance.

From parse doc:
Matches given rule zero or more times (Kleene star), stops if the match failed or if input did not advance. Always succeeds.


Hm. Ok, but how exactly do you *repeat rule zero or more times?* :) 
If you repeat something, it means you did it *at least twice*. That's common-sense definition of repetition. 
But if you repeat something *zero times*, it could mean you
didn't apply the rule at all, or that you did apply it only once. Both would be correct. 

Now when you are learning this, it could be pretty confusing. The way I try to think about all this is:
you have building blocks. Each one of them has well-defined behaviour. And then you just compose them.

I'm all for more meaningful terminology that's understandable not only to developers (although some of them would also find
this kind of terminology *very* confusing). 

So, after analyzing this for some time, I guess this would correctly describe `any `'s behavior?


    1. apply the rule (it either matches, or it doesn't. There is no zero.) 
     
       A) If it doesn't match, it will succeed (return true I suppose).    
   
       B) If it does match: 
            - if there was input advancement, apply the rule again (step 1.)
            - if there wasn't input advancement, it will succeed.
     
     Any is just repetition operator - it never advances input on its own.

**Reply by hiiamboris** [reply]
Yes, that's a correct description of its behavior. As for the natural language limitations, even wikipedia uses the ["zero or more repetitions" phrasing](https://en.wikipedia.org/wiki/Kleene_star) so I guess nobody has yet found a phrasing better understood by humans.

**Reply by gor77** [reply]
The fact that something is on Wikipedia in a specific form doesn't really mean much. People react, or they don't, they fight,
and then things somehow end. Until the next time. And that's it. 
Sometimes the end result is very good. Sometimes it's awful. 

After all, maybe people were confused many times by what's written there, you can't know that for sure
if they didn't react. Or if nobody listened to them. 

It's all about understanding. 
And if the aim of the language really is reducing complexity, sometimes it's good to get a fresh perspective on things. 

No matter how well established these things are. 

I'll rest my case here :)

**Reply by greggirwin** [reply]
Here's how it's worded in the docs (https://github.com/red/docs/blob/master/en/parse.adoc#any):

"Matches given rule zero or more times ([Kleene star](https://en.wikipedia.org/wiki/Kleene_star)), stops if the match failed or if input did not advance. Always succeeds."

Seems about as clear as it can be, to me.

**Reply by gor77** [reply]
I think that's the point :) Maybe people forget how this looks like to newcomers when they read this.

For example, at first I couldn't distinguish "match" from "success" and I thought of them as synonyms. Especially after I read 
 [core-principles](https://github.com/red/docs/blob/master/en/parse.adoc#12-core-principles):

    When a given rule matches, it succeeds, and Parse (optionally) advances ...

    If there is no match, rule fails, ...


Maybe I've missed it but I couldn't find anything else about match and success and their exact meaning.

In the case of `any`, the result is success, even if the rule didn't match. 

Then there was confusion about "_matches given rule zero or more times_".
How can a repetition operator match any given rule zero times? 
After lots of head bashing, and Boris explanation, I realized it's really not about matching, it's about results: 
`Any `will succeed even if there was no match.  It will apply the rule, and if there is no match it will actually succeed.
Which seems to be in contradiction with the core principles.

At least to me, there is a huge difference when you say:

1) _matches a given rule zero times_ vs.
2) _apply the given rule and if there is no match, return success_.

Maybe it's the language barrier, I don't know. 

But the second line is something I understand.
The first one I probably never will :)
